src/storage/secondary/HttpStorage.cpp

```diff
@@ -0,0 +1,218 @@
+// Copyright (C) 2021 Joel Rosdahl and other contributors
+//
+// See doc/AUTHORS.adoc for a complete list of contributors.
+//
+// This program is free software; you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program; if not, write to the Free Software Foundation, Inc., 51
+// Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+
+#include "HttpStorage.hpp"
+
+#include <Digest.hpp>
+#include <Logging.hpp>
+#include <ccache.hpp>
+#include <exceptions.hpp>
+#include <fmtmacros.hpp>
+
+#include <third_party/httplib.h>
+#include <third_party/nonstd/string_view.hpp>
+
+namespace storage {
+namespace secondary {
+
+namespace {
+
+nonstd::string_view
+to_string(httplib::Error error)
+{
+  using httplib::Error;
+
+  switch (error) {
+  case Error::Success:
+    return "Success";
+  case Error::Connection:
+    return "Connection";
+  case Error::BindIPAddress:
+    return "BindIPAddress";
+  case Error::Read:
+    return "Read";
+  case Error::Write:
+    return "Write";
+  case Error::ExceedRedirectCount:
+    return "ExceedRedirectCount";
+  case Error::Canceled:
+    return "Canceled";
+  case Error::SSLConnection:
+    return "SSLConnection";
+  case Error::SSLLoadingCerts:
+    return "SSLLoadingCerts";
+  case Error::SSLServerVerification:
+    return "SSLServerVerification";
+  case Error::UnsupportedMultipartBoundaryChars:
+    return "UnsupportedMultipartBoundaryChars";
+  case Error::Compression:
+    return "Compression";
+  case Error::Unknown:
+    break;
+  }
+
+  return "Unknown";
+}
+
+int
+get_url_port(const Url& url)
+{
+  if (!url.port().empty()) {
+    size_t idx = 0;
+    const auto port = std::stoi(url.port(), &idx);
+    if (idx != url.port().size()) {
+      throw Error("Cannot convert port to integer: {}", url.port());
+    }
+    return port;
+  }
+  if (url.scheme() == "http") {
+    return 80;
+  } else {
+    throw Error("Unknown scheme: {}", url.scheme());
+  }
+}
+
+std::string
+get_url_path(const Url& url)
+{
+  auto path = url.path();
+  if (path.empty() || path.back() != '/') {
+    path += '/';
+  }
+  return path;
+}
+
+} // anonymous namespace
```

Comment:
For foolish consistency with other code:

```suggestion
} // namespace
```

(That is the comment that clang-format adds by default.)